third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1221,4 +1582,247 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+      }
+
+      outer_reduction_tv->split(
+          axisID, NamedScalar::getParallelDim(ParallelType::BIDy));
+      outer_reduction_tv->axis(axisID--)->parallelize(ParallelType::BIDy);
+    }
+
+    auto outer_reference_tv =
+        reduction_scheduler_utils::sortAndRFactor(outer_reduction_tv);
+    outer_reference_tvs.emplace_back(outer_reference_tv);
+  }
+}
+
+void schedulePersistentKernelInnerOuter(
+    Fusion* fusion,
+    const ReductionParams& rparams) {
+  FUSER_PERF_SCOPE("schedulePersistentKernelInnerOuter");
+
+  FusionGuard fg(fusion);
+
+  // Grab the reduction, input, and output tensor views. dummy_outputs are
+  // helper tensors for persistent buffer projection.
+  std::vector<TensorView*> dummy_outputs, cached_inputs, reduction_tvs;
+  std::vector<std::pair<TensorView*, TensorView*>> cached_outputs;
+  beforeSchedule(
+      fusion,
+      rparams,
+      dummy_outputs,
+      cached_inputs,
+      reduction_tvs,
+      cached_outputs);
+
+  // split reduction_tvs into inner and outer reduction_tvs
+  std::vector<TensorView*> inner_reduction_tvs, outer_reduction_tvs;
+  for (auto tv : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(tv)) {
+      inner_reduction_tvs.emplace_back(tv);
+    } else {
+      outer_reduction_tvs.emplace_back(tv);
+    }
+  }
+  TORCH_INTERNAL_ASSERT(
+      inner_reduction_tvs.size(),
+      "schedulePersistentKernelInnerOuter is called but no inner reduction is found.");
+  TORCH_INTERNAL_ASSERT(
+      outer_reduction_tvs.size(),
+      "schedulePersistentKernelInnerOuter is called but no outer reduction is found.");
+
+  // schedule inner reduction, only schedule the first inner reduction tv, then
+  // will be propagated to other inner reduction tvs.
+  TensorView* inner_reference_tv =
+      scheduleReductionGeneral(fusion, rparams, inner_reduction_tvs);
+
+  // schedule outer reduction, schedule all the outer reduction tvs since we
+  // need to store the intermediate results.
+  std::vector<TensorView*> cached_gmem;
+  std::vector<TensorView*> cached_gmem_reload;
+  std::vector<TensorView*> outer_reference_tvs;
+  std::unordered_set<TensorView*> boundaryNodesSet;
+  scheduleReductionCombinedOuter(
+      fusion,
+      rparams,
+      outer_reduction_tvs,
+      cached_gmem,
+      cached_gmem_reload,
+      outer_reference_tvs,
+      boundaryNodesSet);
+
+  // Propagate inner reduction and outer reductions
+  for (auto output : dummy_outputs) {
+    fusion->addOutput(output);
+  }
+
+  const bool unroll = rparams.isUnrolled();
+  const bool vectorize =
+      rparams.vectorize_inner_reduction || rparams.vectorize_iter_dom;
+  const bool is_outer_grid_persistence = rparams.persistent_kernel &&
+      rparams.cross_grid_inner_reduction && !rparams.fastest_dim;
+
+  // Propagate inner reduction. There is a cutoff at boundaryNodesSet, so this
+  // propagation will not propagate to the outer reduction.
```

Comment:
Is this the only place `boundaryNodeSet` gets tensors?

https://github.com/csarofeen/pytorch/pull/2400/files#diff-7deb2b8b109b1bbc59af5ef572bb4786bd7eb0188fbf29110a4af1f279cf1f97R1625

If so, doesn't the following propagation include the first stage of the outer reduction? If  `partialResultReloaded` is the boundary, isn't `partialResult` reachable from `inner_reference_tv`?

Reply:
yes. `partialResult` follows the pattern of `inner_reference_tv`. It was calculated with inner reduction.

Reply:
So, the comment isn't quite right? It does propagate to part of the outer reduction expressions, right?

Reply:
oh, yes. I was supposed to say "final outer reduction". The partial outer reduction was computed with inner reduciton.

Reply:
Have you updated the comment?

Reply:
revised. don't know why not updated here.